### DISCUSSION TITLE
Upgrading from `0.50.0` to `0.52.3`

### DISCUSSION THREAD

**Discussion by MitchTurner** [libp2p/rust-libp2p#4674] [Q&A]
We are currently upgrading our `libp2p` version and I thought I would start a discussion here for any ongoing questions we have in the process. LMK if there would be a better way of organizing our questions. 

It is possible that our architecture is also out of date/wrong. Hopefully this will also be an opportunity to update the architecture as well.

***

I'm not sure of all the issues we are going to have; we have already successfully converted a lot of the code. But I'll start with the current problem:

### Background

We have a custom network behavior `PeerReportBehavior` which `on_swarm_event`, will take the `peer_id` from a `FromSwarm::ConnectionEstablished` and lookup the addresses of that `peer_id` using `addresses_of_peer` which under the hood is:
```rs
    fn addresses_of_peer(&mut self, peer_id: &PeerId) -> Vec<Multiaddr> {
        self.identify.addresses_of_peer(peer_id)
    }
```

Obviously, `addresses_of_peer` is no longer available from `PeerReportBehavior` or `identify::Behavior`, so I thought maybe `identify::Behavior` might expose a getter for `discovered_peers` which is what the old `addresses_of_peer` used to do [here](https://github.com/libp2p/rust-libp2p/blob/v0.50.0/protocols/identify/src/behaviour.rs#L403-L405). It doesn't.

### Question

What is the correct way of getting those peer addresses? 

I could just hit `handle_pending_outbound_connections` which will return the desired addresses as seen [here](https://github.com/libp2p/rust-libp2p/blob/43575e9918bd2c5d0dea076d52f82d4f31b42bb1/protocols/identify/src/behaviour.rs#L327-L340). It just has a bunch of additional (albeit unused in this case) parameters, which makes me feel like this isn't the intended use. 

I would appreciate your perspective and guidance avoiding any pitfalls here :)

**Comment by thomaseizinger** [comment]
> We are currently upgrading our `libp2p` version and I thought I would start a discussion here for any ongoing questions we have in the process. LMK if there would be a better way of organizing our questions.

All good! This is what discussions are for :)

> I could just hit `handle_pending_outbound_connections` which will return the desired addresses as seen [here](https://github.com/libp2p/rust-libp2p/blob/43575e9918bd2c5d0dea076d52f82d4f31b42bb1/protocols/identify/src/behaviour.rs#L327-L340). It just has a bunch of additional (albeit unused in this case) parameters, which makes me feel like this isn't the intended use.

Correct. All APIs on `NetworkBehaviou` are "internal" in that, you should never call them yourself unless you are wrapping another `NetworkBehaviour` and are just forwarding calls.

There is an open issue and ongoing effort about sharing peer addresses across `NetworkBehaviour`s: https://github.com/libp2p/rust-libp2p/issues/4302. That will make it possible to easily implement a [`PeerStore`](https://github.com/libp2p/rust-libp2p/issues/4103) which can gather and save addresses of peers that are _discovered_ by other peers.

For the moment, it is unfortunately only possible to be notified about the addresses you get connected to.

What function does your `PeerReportBehaviour` fulfill?

**Reply by MitchTurner** [reply]
> Correct. All APIs on NetworkBehaviou are "internal" in that, you should never call them yourself unless you are wrapping another NetworkBehaviour and are just forwarding calls.

Does this not qualify as "wrapping another NetworkBehavioer", as we are wrapping the `identify::Behavior`? Or are you specifically referring to the sort wrapping that happens in the `#[derive(NetworkBehaviour)]` for `poll()`, etc? I guess that distinction is blurry for me still.

> There is an open issue and ongoing effort about sharing peer addresses across NetworkBehaviours: https://github.com/libp2p/rust-libp2p/issues/4302. That will make it possible to easily implement a https://github.com/libp2p/rust-libp2p/issues/4103 which can gather and save addresses of peers that are discovered by other peers.

Oh nice! Maybe we can weigh in!

> For the moment, it is unfortunately only possible to be notified about the addresses you get connected to.

Hmm. Hopefully that isn't a hard blocker for us...

> What function does your PeerReportBehaviour fulfill?

It is a union of the `identity::Behavior` and our own `HeartBeat` behavior. It forwards any events on to those sub-behaviors, and it also forwards information to our service to manage all our peers from a domain-specific perspective, i.e. when was its last heartbeat, and what is it's score as a peer, etc. [Here](https://github.com/FuelLabs/fuel-core/blob/67042b0032c2e9c062835e3038475e2549b6a524/crates/services/p2p/src/peer_report.rs#L127) is a link to that `PeerReportBehavior` just for reference.

I'm still trying to make sense of a lot of this code, as I've recently inherited it. And in formulating an answer for you, I've been reading through what consumes the `addresses` and it might be possible that we don't even need them. So I hope this hasn't been a waste of your time :). I'm going to go down that route for a while and perhaps I can move onto the next upgrade issue to bug you with :+1:

**Reply by thomaseizinger** [reply]
> > Correct. All APIs on NetworkBehaviou are "internal" in that, you should never call them yourself unless you are wrapping another NetworkBehaviour and are just forwarding calls.
> 
> Does this not qualify as "wrapping another NetworkBehavioer", as we are wrapping the `identify::Behavior`?

What you are doing does qualify as that :)

Skimming the code, I think you might be better served by ditching your `PeerReportBehaviour` completely and instead using `#[derive(NetworkBehaviour)]`!

An event loop like it is written in https://github.com/libp2p/rust-libp2p/blob/82902fc9f9a9a85d38ee62fdd11d3cd7bf8f5474/examples/file-sharing/src/network.rs#L170-L403 should allow you to fairly effortlessly combine the state of `identify::Behaviour` and `Heartbeat` and make decisions based on the events of both.

I see that you have something similar in https://github.com/FuelLabs/fuel-core/blob/67042b0032c2e9c062835e3038475e2549b6a524/crates/services/p2p/src/p2p_service.rs#L427-L450. What I'd suggest is:

- Move `identify::Behaviour` and `Heartbeat` into `FuelBehaviour`
- Update `handle_behaviour_event` to make use if the data in `identify::Event` and the output of your `Heartbeat` behaviour to make whatever decisions you need to make. For example, `identify::Event` will contain the self-advertised addresses of each peer.

**Reply by MitchTurner** [reply]
> Move identify::Behaviour and Heartbeat into FuelBehaviour

> Update handle_behaviour_event to make use if the data in identify::Event and the output of your Heartbeat behaviour to make whatever decisions you need to make. For example, identify::Event will contain the self-advertised addresses of each peer.

Yes. I have a feeling you might be right. I just want to make sure that the events are emitting what we need. I have a sneaking suspicion that our `Heartbeat` behavior, as well as another `Discovery` behavior might be obsolete as well, which would simplify our code a lot. It's just not ideal ripping things out at the same time we're upgrading, so trying to triage as best as I can.

I'm going to keep digging so I can talk more authoritatively about how this _should_ work.

**Reply by thomaseizinger** [reply]
Do you know about `libp2p-swarm-test`? We use it internally to test all our behaviours: https://github.com/libp2p/rust-libp2p/blob/master/swarm-test/src/lib.rs

Makes it really easy to spin up several `Swarm`s that are connected via memory-addresses. This is how we test the events of our network behaviours. See for example: https://github.com/libp2p/rust-libp2p/blob/82902fc9f9a9a85d38ee62fdd11d3cd7bf8f5474/protocols/identify/tests/smoke.rs#L81-L129

**Reply by thomaseizinger** [reply]
There is a lot of type-inference magic going on though so let me know if you have any questions on how to use it :sweat_smile:

**Reply by MitchTurner** [reply]
> Do you know about libp2p-swarm-test?

Yes! We had a [nice conversation](https://github.com/libp2p/rust-libp2p/pull/4508) about it a few weeks ago (I think I've changed my icon since then :)).

That's part of the reason I'm pushing for this upgrade: I'm not happy about the state of the tests rn and I'm hoping we can get some better coverage and maintainability. `libp2p-swarm-test` wasn't supported at `0.50.0` IIRC.

**Reply by thomaseizinger** [reply]
> > Do you know about libp2p-swarm-test?
> 
> Yes! We had a [nice conversation](https://github.com/libp2p/rust-libp2p/pull/4508) about it a few weeks ago (I think I've changed my icon since then :)).

Ah sorry about not recognizing you! Your nick was familiar but I didn't make the connection back to that thread :)

**Comment by mxinden** [comment]
Closing here. Seems resolved. Please let us know if you have more questions.

**Reply by MitchTurner** [reply]
We had to put the work on pause. Will be returning to it as soon as possible.
